### PR TITLE
v5: use react-native-pager-view and simplify code

### LINKED ISSUE #159: how to add Icons for  tabs
hi, i need to ask that how can we use icons for tabs, now we are just using label.

### PR DESCRIPTION
**Pull request by andreialecu**
This commit adds a peer dependency on `react-native-pager-view`.

You will need to install `react-native-pager-view@5` into your project if you don't have it already.

The purpose is to fix some issues that seem to have appeared on Android with recent react-native and reanimated versions.

In addition it also adds a property to `MaterialTabBar` that allows keeping the currently active tab always centered when using a scrollable tab bar.

Closes #159

### CODE REVIEW

File: example/package.json [hunk 1]
```diff
@@ -11,7 +11,8 @@
     "react-dom": "16.13.1",
     "react-native": "https://github.com/expo/react-native/archive/sdk-42.0.0.tar.gz",
     "react-native-gesture-handler": "~1.10.2",
-    "react-native-reanimated": "~2.2.0",
+    "react-native-pager-view": "^5",
```
Comment (PedroBern) [inline example/package.json:14]: ❤️

**Comment by andreialecu** [comment]
There's a known issue about `react-native-gesture-handler` buttons now allowing to swipe panes, see #256

**Comment by alexco2** [comment]
In my production app, I saw two errors with the workaround. Scrollview would not snap anymore and navigating between tabs is not possible if a header is in an unsnapped position. Scrolling between tabs works though. Does this pr fix the problems introduced with reanimated? Do you need help testing?

**Comment by alexco2** [comment]
So I tested it. The `Scrollview` still does not snap. Could it be that you deactivated it? The bug that pressing a `TabBar` would not change tabs if the header is not snapped is fixed though. One thing I saw is that if you wrap `Tabs.Container` inside a `View`, the layout is broken.

**Comment by andreialecu** [comment]
Both iOS and Android? 

Regarding the wrapping view. You may need to pass a `style` of `flex: 1` to it

**Comment by alexco2** [comment]
> Both iOS and Android?
> 
> Regarding the wrapping view. You may need to pass a `style` of `flex: 1` to it

yes, I realised that and edited my answer. I found a bug with snapThreshold though. Can you confirm it.
I am currently just testing in Android.

**Comment by alexco2** [comment]
@andreialecu 
I came across another Bug, which I actually already reported in `react-native-pager-view`. (https://github.com/callstack/react-native-pager-view/issues/440)
When navigating back from a pager view, the app sometimes crashes with the below error code, because the pages are not recycled.  I remember that I moved to `react-native-vollapsilbe-tab-view` since this Bug was not being resolved.
Apparently `react-native-pager-view` version 6 solves this issue. I tried it with react-native-collapsible-tab-view but the Version is not compatible. I get the following errors.

`TypeError: Cannot read property 'displayName' of undefined, js engine: hermes`

`Error: Requiring module "node_modules/react-native-collapsible-tab-view/src/index.tsx", which threw an exception: TypeError: Cannot read property 'displayName' of undefined, js engine: hermes`

`TypeError: Cannot read property 'Tabs' of undefined`

Here the initial Error when using version 5 of `react-native-pager-view`
```
Fatal Exception: java.lang.IllegalArgumentException: Scrapped or attached views may not be recycled. isScrap:false isAttached:true androidx.viewpager2.widget.f$m{9e238b7 VFED..... ......ID 0,0-1080,2096 #9}, adapter:com.reactnativepagerview.g@7241824, layout:androidx.viewpager2.widget.f$h@9916f8d, context:com.facebook.react.uimanager.p0@333f20
       at androidx.recyclerview.widget.RecyclerView$v.C(:282)
       at androidx.recyclerview.widget.RecyclerView.removeAnimatingView(:22)
       at androidx.recyclerview.widget.RecyclerView$m.a(:27)
       at androidx.recyclerview.widget.RecyclerView$l.h(:7)
       at androidx.recyclerview.widget.n.G(:3)
       at androidx.recyclerview.widget.c$d.onAnimationEnd(:17)
       at android.view.ViewPropertyAnimator$AnimatorEventListener.onAnimationEnd(ViewPropertyAnimator.java:1112)
       at android.animation.Animator$AnimatorListener.onAnimationEnd(Animator.java:555)
       at android.animation.ValueAnimator.endAnimation(ValueAnimator.java:1250)
       at android.animation.ValueAnimator.doAnimationFrame(ValueAnimator.java:1492)
       at android.animation.AnimationHandler.doAnimationFrame(AnimationHandler.java:146)
       at android.animation.AnimationHandler.access$100(AnimationHandler.java:37)
       at android.animation.AnimationHandler$1.doFrame(AnimationHandler.java:54)
       at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1106)
       at android.view.Choreographer.doCallbacks(Choreographer.java:866)
       at android.view.Choreographer.doFrame(Choreographer.java:792)
       at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1092)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loopOnce(Looper.java:226)
       at android.os.Looper.loop(Looper.java:313)
       at android.app.ActivityThread.main(ActivityThread.java:8663)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:567)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1135)
```

**Comment by andreialecu** [comment]
You can try v6 and it should work if you update the pager view import to a named import instead of the default import.

**Comment by alexco2** [comment]
@andreialecu 
Thank you for the tip. That makes version 6 compatible, but navigating to and quickly back from a page with pager-view unfortunately still crashes the app. Apparently, the issue is not solved with version 6 of `react-native-pager-view`.
How do you evaluate this problem?

**Comment by andreialecu** [comment]
I didn't notice that crash but you could try applying https://github.com/callstack/react-native-pager-view/pull/552

**Comment by alexco2** [comment]
@andreialecu 
Wow, the PR solves this bug. Hopefully it gets merged soon. If now,  I will try to publish a npm package with a the patch.

**Comment by andreialecu** [comment]
> The `Scrollview` still does not snap. Could it be that you deactivated it? Adding `snapThreshold` to a container with `Scrollviews` breaks the synchronisation of the header between tab changes.

@alexco2 can you provide more info on this? In the example app snapping seems to work properly for scrollviews.

**Comment by andreialecu** [comment]
Released `5.0.0-rc.6` with some improvements and bug fixes. 

The snapping now uses a reanimated reaction for more control, which should hopefully prevent some race conditions and desynchronizations.

Some feedback on it would be great.

**Comment by andreialecu** [comment]
I noticed a bug where the header may sometimes start collapsed 🤔

**Comment by alexco2** [comment]
@andreialecu I will get back to you later today or latest tomorrow

**Comment by alexco2** [comment]
> Released `5.0.0-rc.6` with some improvements and bug fixes. 
> 
> The snapping now uses a reanimated reaction for more control, which should hopefully prevent some race conditions and desynchronizations.
> 
> Some feedback on it would be great.
> 

I will test it in my app and report asap

**Comment by alexco2** [comment]
> > The `Scrollview` still does not snap. Could it be that you deactivated it? Adding `snapThreshold` to a container with `Scrollviews` breaks the synchronisation of the header between tab changes.
> 
> @alexco2 can you provide more info on this? In the example app snapping seems to work properly for scrollviews.

I tried it with rc.6 and could not reproduce it :)

**Comment by alexco2** [comment]
> Released `5.0.0-rc.6` with some improvements and bug fixes.
> 
> The snapping now uses a reanimated reaction for more control, which should hopefully prevent some race conditions and desynchronizations.
> 
> Some feedback on it would be great.

Some stuff I noticed:

1.  when navigating from a tab with collapsed header to a tab where the header was previously not collapsed, the content seems to be scrolled up more than necessary
2. Sometimes I manage to get the header "Stuck" when navigating between tabs, but this happens really rarely. I don't know how to reproduce it unfortunately

Other than that the snapping works really good :)

**Comment by alexco2** [comment]
Also, `Tabs.Scrollview` has the following typescript error. Do you have that too?

`Type '{ children: Element[]; overScrollMode: "never"; style: { transform: { translateY: number; }[]; }; contentContainerStyle: { justifyContent: "center"; alignItems: "flex-start"; flexDirection: "row"; transform: { ...; }[]; }; scrollEventThrottle: number; showsVerticalScrollIndicator: false; }' is missing the following properties from type 'Pick<ScrollViewProps, "testID" | "style" | "scrollEnabled" | "keyboardDismissMode" | "onMoveShouldSetResponderCapture" | "onStartShouldSetResponder" | "overScrollMode" | ... 98 more ... | "persistentScrollbar">': accessibilityComponentType, accessibilityTraits`

**Comment by alexco2** [comment]
@andreialecu this pr might also be a practicable way to upgrade to rea 3.0
Apparently there are no real breaking changes. (No more rea 1.0, compile from source, fabric enabled)
https://t.co/GNWy3AzfqE

**Comment by andreialecu** [comment]
The changes in this PR specify a peer dependency range of `"react-native-reanimated": ">=2.3.0"` so if there are no breaking changes in v3 as they claim, it should already just work.

Feel free to test it and let me know how it went.

**Comment by alexco2** [comment]
Oh good to know. Will test it today or tomorrow.

**Comment by azwinlam** [comment]
Sorry, could I have some pointers on how to try out the PR? Currently on     "react-native-collapsible-tab-view": "^5.0.0-rc.0", and it's crashing on expo publish (apk + testflight both platforms)

**Comment by andreialecu** [comment]
Make sure you update reanimated. 2.2.x is not supported.

Also ensure the peer dependency in the top post is met.

What is the crash?

**Comment by azwinlam** [comment]
Let me try to run minify and production mode on expo, but it's on APK so the crash is not logged.
    "react-native-collapsible-tab-view": "^5.0.0-rc.0",
    "react-native-pager-view": "^5.4.15",
    "react-native-reanimated": "~2.8.0",

**Comment by alexco2** [comment]
@azwinlam have you tried rc.6?

**Comment by alexco2** [comment]
@andreialecu I´m getting some build error. I suggest waiting until the release is finished :D

**Comment by azwinlam** [comment]
rc9 fixes the tab button not working in previous 4.5.2 but this only works in the expo dev setting. The app will crash on any screen with the collapsible tab on publishing apk/ios.

**Comment by andreialecu** [comment]
Please share the crash logs, the feedback is not actionable otherwise.

**Comment by azwinlam** [comment]
It's crashing on production APK, I currently do not how to log crashes on production builds. Will update if I find something.

**Comment by alexco2** [comment]
@azwinlam I can recommend freebase crashlytics. If you use firebase, this has minimal setup effort

**Comment by andreialecu** [comment]
I'm going to merge this to `master`, but will not release it to NPM yet as stable.

It will still be on `@rc` until everything is confirmed to be working right. I'm using it in some projects without issues so far, but will wait for further confirmation.

**Comment by azwinlam** [comment]
@andreialecu I've looked around and tried using ADB logcat to log the crash. I might have found it.

https://ibb.co/rvZ4fhm

`05-26 10:36:33.264 14641 14665 E AndroidRuntime: FATAL EXCEPTION: expo-updates-error-recovery
05-26 10:36:33.264 14641 14665 E AndroidRuntime: Process: com.cob2.app, PID: 14641
05-26 10:36:33.264 14641 14665 E AndroidRuntime: com.facebook.react.common.JavascriptException: Invariant Violation: requireNativeComponent: "RNCViewPager" was not found in the UIManager.
05-26 10:36:33.264 14641 14665 E AndroidRuntime:
05-26 10:36:33.264 14641 14665 E AndroidRuntime: This error is located at:
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in RNCViewPager
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in w
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in AnimatedComponent(w)
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in RCTView
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in AnimatedComponent(View)
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in C
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in RCTView
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in RCTView
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in RNSScreen
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in y
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in L
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in f
05-26 10:36:33.264 14641 14665 E AndroidRuntime:     in Unknown`

**Comment by andreialecu** [comment]
By that error it seems you don't have `react-native-pager-view` installed, which is now a prerequisite as mentioned in the top comment.

**Comment by azwinlam** [comment]
Thanks, it was my oversight. It appears to be working and tabs work as expected. Thanks @andreialecu

**Pull request merged**